File: M-33-search-in-rotated-sorted-array/solution.py
class Solution:

    def binary_search(self, arr, target):
        l, r = 0, len(arr)
        while l < r:
            m = l + (r - l) // 2
            if (arr[m] > arr[0] and target < arr[m]) or (arr[m] < arr[-1] and not arr[m] <= target < arr[-1]):
                r = m
            else:
                l = m + 1

        return l

    def search(self, nums, target):
        pass


sol = Solution()

print(sol.binary_search([4, 5, 6, 7, 0, 1, 2], 2))
print(sol.binary_search([7, 1, 2, 3, 4, 5, 6], 5))
print(sol.binary_search([4, 5, 6, 7, 8, 0, 1, 2], 4))

File: M-33-search-in-rotated-sorted-array/solution.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function (nums, target) {
  let left = 0;
  let right = nums.length - 1;

  while (left < right) {
    let mid = left + Math.floor((right - left) / 2);

    if (nums[mid] > nums[right]) {
      if (target >= nums[left] && target <= nums[mid]) {
        right = mid;
      } else {
        left = mid + 1;
      }
    } else {
      if (target > nums[mid] && target <= nums[right]) {
        left = mid + 1;
      } else {
        right = mid;
      }
    }
  }

  return nums[left] === target ? left : -1;
};

// console.log(search([4, 5, 6, 7, 0, 1, 2], 0)); // 4
// console.log(search([0, 1, 3, 4, 5, 6, 7], 0)); // 0
console.log(search([6, 7, 0, 1, 2, 4, 5], 4)); // 0
// console.log(search([4, 5, 6, 7, 0, 1, 2], 3)); // -1
// console.log(search([1], 0)); // -1
